File: skills/core/scripts/validate.py
#!/usr/bin/env python3
"""
CORE v2 validation runner.
Usage: python3 validate.py <project-path>

Reads test corpus from <project-path>/_memories/_validation/tests/test-*.yaml
Runs Tier 1 retrieval simulation (grep) for each test
Scores precision + recall against expected/forbidden unit lists
Writes report to <project-path>/_outputs/validation/<date>/REPORT.md

Layout per DC-68: units are FLAT at <project>/_memories/, not in type-subdirectories.
"""
import sys, re
from pathlib import Path
from datetime import datetime

def parse_frontmatter(content):
    """Parse YAML frontmatter from a markdown file content.
    Uses PyYAML if available, otherwise a fallback parser that handles
    multi-line lists.
    """
    match = re.match(r'^---\n(.*?)\n---', content, re.DOTALL)
    if not match:
        return {}
    fm_text = match.group(1)
    try:
        import yaml
        return yaml.safe_load(fm_text) or {}
    except ImportError:
        pass
    # Fallback: handle scalars + multi-line lists
    result = {}
    current_key = None
    current_list = None
    for line in fm_text.split('\n'):
        if not line.strip():
            continue
        # List continuation
        if line.startswith('  - ') or line.startswith('- '):
            value = line.lstrip()[2:].strip().strip('"').strip("'")
            if current_list is not None:
                current_list.append(value)
            continue
        # Top-level key
        if ':' in line and not line.startswith(' '):
            if current_key and current_list is not None:
                result[current_key] = current_list
                current_list = None
            k, _, v = line.partition(':')
            k = k.strip()
            v = v.strip()
            current_key = k
            if not v:
                current_list = []
            else:
                if v.startswith('[') and v.endswith(']'):
                    result[k] = [x.strip().strip('"').strip("'") for x in v[1:-1].split(',') if x.strip()]
                else:
                    result[k] = v.strip('"').strip("'")
                current_key = None
                current_list = None
    if current_key and current_list is not None:
        result[current_key] = current_list
    return result

def load_tests(project_path):
    tests_dir = Path(project_path) / "_memories/_validation/tests"
    tests = []
    if not tests_dir.exists():
        return tests
    for f in sorted(tests_dir.glob("test-*.yaml")):
        content = f.read_text()
        fm = parse_frontmatter(content)
        if fm and 'query' in fm:
            tests.append(fm)
    return tests

STOPWORDS = {
    # Length-3 common words that survive length>=3 filter but carry no signal
    "the", "and", "for", "are", "was", "but", "not", "you", "all", "any",
    "can", "had", "has", "his", "her", "how", "its", "may", "now", "one",
    "our", "out", "own", "see", "she", "two", "use", "via", "way", "who",
    "why", "yes", "yet",
    # Length-4+ common words
    "what", "where", "which", "would", "could", "should", "their", "there",
    "these", "those", "about", "after", "again", "before", "being", "below",
    "between", "during", "other", "while", "every", "based", "into", "than",
    "then", "this", "that", "they", "from", "have", "with", "your", "just",
    "like", "make", "more", "much", "only", "over", "some", "such", "very",
    "when", "will", "also", "back", "been", "both", "down", "even", "ever",
    "here", "many", "much", "must", "need", "same", "well",
    # Negations and contractions reduced after punctuation strip
    "dont", "isnt", "wont", "cant", "doesnt", "wouldnt", "couldnt",
}

def simulate_retrieval_tier1(query, project_path, top_k=3):
    """Tier 1 retrieval simulator with term-density ranking.

    Approach (improved per Phase 7 Fault F1 + ANALYSIS.md):
    1. Extract candidate terms from query (length > 4, not in STOPWORDS).
    2. For each unit file, count distinct terms matched in body + filename.
       Filename matches count 2x (filename slug is a strong intent signal).
    3. Rank candidates by score descending; return top-K.
    4. Floor: any unit with zero matches is dropped regardless of K.

    top_k is scaled by the caller to match `len(expected_memories) + 1`
    (test-by-test) so precision can actually reach 0.8 when the expected
    unit is the top-ranked match. Default 3 if not scaled.

    This replaces the v1 OR-of-terms approach that surfaced every unit
    containing any term — structurally noisy on a 60-unit corpus.

    Flat layout per DC-68 — walks _memories/*.md (recursive into archive
    subfolders is excluded).
    """
    units_dir = Path(project_path) / "_memories"
    raw_terms = [t.lower().strip(".,?!()[]\"'") for t in query.split()]
    # length >= 3 keeps acronyms (MCP, v2) and short content words (flat,
    # unit, store, code) while STOPWORDS filters out the common noise.
    terms = [t for t in raw_terms if len(t) >= 3 and t not in STOPWORDS]
    if not terms:
        return []

    scored = []
    for f in units_dir.rglob("*.md"):
        path_str = str(f)
        if "archive" in path_str or "cold-storage" in path_str or "_validation" in path_str:
            continue
        if f.name.startswith("INDEX") or f.name.startswith("README") or f.name.startswith("_"):
            continue
        try:
            content = f.read_text().lower()
            slug = f.stem.lower()
            body_hits = sum(1 for t in terms if t in content)
            slug_hits = sum(1 for t in terms if t in slug)
            score = body_hits + 2 * slug_hits
            if score > 0:
                scored.append((score, f.stem))
        except Exception:
            pass

    scored.sort(key=lambda x: (-x[0], x[1]))
    return [stem for _, stem in scored[:top_k]]

def score_precision_recall(retrieved, expected, forbidden):
    ret_set = set(retrieved)
    exp_set = set(expected) if expected else set()
    forb_set = set(forbidden) if forbidden else set()

    clean_ret = ret_set - forb_set
    tp = len(clean_ret & exp_set)
    fp = len(ret_set & forb_set)
    fn = len(exp_set - ret_set)

    if not exp_set:
        precision = 0.0 if fp > 0 else 1.0
        recall = 1.0
    else:
        precision = tp / max(len(clean_ret), 1)
        recall = tp / max(len(exp_set), 1)
    return round(precision, 2), round(recall, 2)

def main():
    if len(sys.argv) < 2:
        print("Usage: python3 validate.py <project-path>")
        sys.exit(1)

    project_path = sys.argv[1]
    tests = load_tests(project_path)

    if not tests:
        print(f"No tests found in {project_path}/_memories/_validation/tests/")
        sys.exit(0)

    results = []
    for t in tests:
        query = t.get('query', '')
        expected = t.get('expected_memories', [])
        forbidden = t.get('forbidden_memories', [])
        if isinstance(expected, str):
            expected = [expected] if expected else []
        if isinstance(forbidden, str):
            forbidden = [forbidden] if forbidden else []

        # Scale top_k to the test's expected-set size. Strict top_k (no
        # noise buffer) means PASS requires the expected unit(s) to be the
        # top-ranked candidates — the meaningful success criterion for a
        # ranked-retrieval simulator. If the simulator places the expected
        # unit at rank 1 for a single-expected test, P=1.0 and R=1.0.
        top_k = max(1, len(expected))
        retrieved = simulate_retrieval_tier1(query, project_path, top_k=top_k)
        p, r = score_precision_recall(retrieved, expected, forbidden)

        if p >= 0.8 and r >= 0.8:
            status = "PASS"
        elif p < 0.5 or r < 0.5:
            status = "FAIL"
        else:
            status = "INVESTIGATE"

        results.append({
            'query': query,
            'precision': p,
            'recall': r,
            'status': status,
            'retrieved': retrieved[:5],
            'expected': expected,
        })
        icon = "PASS" if status == "PASS" else ("FAIL" if status == "FAIL" else "INV")
        print(f"[{icon}] P={p} R={r} -- {query[:60]}")

    passes = sum(1 for r in results if r['status'] == 'PASS')
    fails = sum(1 for r in results if r['status'] == 'FAIL')
    total = len(results)

    date_str = datetime.now().strftime("%Y-%m-%d")
    report_dir = Path(project_path) / f"_outputs/validation/{date_str}"
    report_dir.mkdir(parents=True, exist_ok=True)

    report_lines = [
        f"# Validation Report -- {date_str}\n\n",
        f"**Results: {passes}/{total} pass, {fails} fail**\n\n",
        "| Status | P | R | Query |\n",
        "|---|---|---|---|\n",
    ]
    for r in results:
        report_lines.append(f"| {r['status']} | {r['precision']} | {r['recall']} | `{r['query'][:70]}` |\n")

    report_lines.append("\n## Detail\n\n")
    for r in results:
        report_lines.append(f"**{r['query']}**  \nExpected: {r['expected']}  \nRetrieved (top 5): {r['retrieved']}\n\n")

    report_path = report_dir / "REPORT.md"
    report_path.write_text(''.join(report_lines))
    print(f"\nReport: {report_path}")

    if fails > 0:
        print(f"\n{fails} FAIL(s) -- PAUSE-AND-SURFACE TRIGGER if any precision or recall < 0.5")
        sys.exit(1)

if __name__ == "__main__":
    main()

File: skills/core/scripts/validate.mjs
/**
 * CORE v2 validation runner.
 *
 * Usage: node validate.mjs <project-path>
 *
 * Reads test corpus from <project-path>/_memories/_validation/tests/test-*.yaml
 * Runs Tier 1 retrieval simulation (term-density scoring) for each test
 * Scores precision + recall against expected/forbidden unit lists
 * Writes report to <project-path>/_outputs/validation/<date>/REPORT.md
 *
 * Per DC-80 the plugin ships Node.js (.mjs) only.
 */

import { readFileSync, writeFileSync, readdirSync, mkdirSync, realpathSync } from 'node:fs';
import { resolve, join } from 'node:path';
import { fileURLToPath } from 'node:url';

export function parseFrontmatter(content) {
  const match = content.match(/^---\n([\s\S]*?)\n---/);
  if (!match) return {};
  const fmText = match[1];
  const result = {};
  let currentKey = null;
  let currentList = null;
  for (const line of fmText.split('\n')) {
    if (!line.trim()) continue;
    if (line.startsWith('  - ') || line.startsWith('- ')) {
      const value = line.trimStart().slice(2).trim().replace(/^["']|["']$/g, '');
      if (currentList !== null) currentList.push(value);
      continue;
    }
    if (line.includes(':') && !line.startsWith(' ')) {
      if (currentKey && currentList !== null) {
        result[currentKey] = currentList;
        currentList = null;
      }
      const colonIdx = line.indexOf(':');
      const k = line.slice(0, colonIdx).trim();
      const v = line.slice(colonIdx + 1).trim();
      currentKey = k;
      if (!v) {
        currentList = [];
      } else if (v.startsWith('[') && v.endsWith(']')) {
        result[k] = v.slice(1, -1).split(',').map(x => x.trim().replace(/^["']|["']$/g, '')).filter(Boolean);
        currentKey = null;
        currentList = null;
      } else {
        result[k] = v.replace(/^["']|["']$/g, '');
        currentKey = null;
        currentList = null;
      }
    }
  }
  if (currentKey && currentList !== null) result[currentKey] = currentList;
  return result;
}

export function loadTests(projectPath) {
  const testsDir = join(projectPath, '_memories/_validation/tests');
  const tests = [];
  let entries;
  try { entries = readdirSync(testsDir).sort(); } catch { return tests; }
  for (const fname of entries) {
    if (!fname.startsWith('test-') || !fname.endsWith('.yaml')) continue;
    const content = readFileSync(join(testsDir, fname), 'utf8');
    const fm = parseFrontmatter(content);
    if (fm && fm.query) tests.push(fm);
  }
  return tests;
}

export const STOPWORDS = new Set([
  'the','and','for','are','was','but','not','you','all','any',
  'can','had','has','his','her','how','its','may','now','one',
  'our','out','own','see','she','two','use','via','way','who',
  'why','yes','yet',
  'what','where','which','would','could','should','their','there',
  'these','those','about','after','again','before','being','below',
  'between','during','other','while','every','based','into','than',
  'then','this','that','they','from','have','with','your','just',
  'like','make','more','much','only','over','some','such','very',
  'when','will','also','back','been','both','down','even','ever',
  'here','many','much','must','need','same','well',
  'dont','isnt','wont','cant','doesnt','wouldnt','couldnt',
]);

export function simulateRetrievalTier1(query, projectPath, topK = 3) {
  const unitsDir = join(projectPath, '_memories');
  const rawTerms = query.split(/\s+/).map(t => t.toLowerCase().replace(/[.,?!()\[\]"']/g, ''));
  const terms = rawTerms.filter(t => t.length >= 3 && !STOPWORDS.has(t));
  if (!terms.length) return [];

  const scored = [];
  function walkDir(dir) {
    let entries;
    try { entries = readdirSync(dir, { withFileTypes: true }); } catch { return; }
    for (const entry of entries) {
      if (entry.isDirectory()) {
        if (['archive', 'cold-storage', '_validation'].includes(entry.name)) continue;
        walkDir(join(dir, entry.name));
        continue;
      }
      if (!entry.name.endsWith('.md')) continue;
      if (entry.name.startsWith('INDEX') || entry.name.startsWith('README') || entry.name.startsWith('_')) continue;
      const fpath = join(dir, entry.name);
      try {
        const content = readFileSync(fpath, 'utf8').toLowerCase();
        const slug = entry.name.replace(/\.md$/, '').toLowerCase();
        const bodyHits = terms.reduce((n, t) => n + (content.includes(t) ? 1 : 0), 0);
        const slugHits = terms.reduce((n, t) => n + (slug.includes(t) ? 1 : 0), 0);
        const s = bodyHits + 2 * slugHits;
        if (s > 0) scored.push([s, entry.name.replace(/\.md$/, '')]);
      } catch {}
    }
  }
  walkDir(unitsDir);

  scored.sort((a, b) => b[0] - a[0] || a[1].localeCompare(b[1]));
  return scored.slice(0, topK).map(x => x[1]);
}

export function scorePrecisionRecall(retrieved, expected, forbidden) {
  const retSet = new Set(retrieved);
  const expSet = new Set(expected || []);
  const forbSet = new Set(forbidden || []);
  const cleanRet = new Set([...retSet].filter(x => !forbSet.has(x)));
  const tp = [...cleanRet].filter(x => expSet.has(x)).length;
  const fp = [...retSet].filter(x => forbSet.has(x)).length;
  if (!expSet.size) {
    return [fp > 0 ? 0.0 : 1.0, 1.0];
  }
  if (fp > 0) {
    return [0.0, Math.round((tp / Math.max(expSet.size, 1)) * 100) / 100];
  }
  const precision = tp / Math.max(cleanRet.size, 1);
  const recall = tp / Math.max(expSet.size, 1);
  return [Math.round(precision * 100) / 100, Math.round(recall * 100) / 100];
}

export function main(argv) {
  if (!argv[0]) {
    process.stderr.write('Usage: node validate.mjs <project-path>\n');
    process.exit(1);
  }

  const projectPath = resolve(argv[0]);
  const tests = loadTests(projectPath);

  if (!tests.length) {
    console.log(`No tests found in ${projectPath}/_memories/_validation/tests/`);
    process.exit(0);
  }

  const results = [];
  for (const t of tests) {
    const query = t.query || '';
    let expected = t.expected_memories || [];
    let forbidden = t.forbidden_memories || [];
    if (typeof expected === 'string') expected = expected ? [expected] : [];
    if (typeof forbidden === 'string') forbidden = forbidden ? [forbidden] : [];
    const scoreK = Math.max(1, expected.length);
    const candidateK = Math.max(5, scoreK);
    const candidates = simulateRetrievalTier1(query, projectPath, candidateK);
    const retrieved = candidates.slice(0, scoreK);
    const forbiddenHits = candidates.filter(x => forbidden.includes(x));
    const [p, r] = scorePrecisionRecall(retrieved, expected, forbidden);
    const status = forbiddenHits.length ? 'FAIL' : (p >= 0.8 && r >= 0.8) ? 'PASS' : (p < 0.5 || r < 0.5) ? 'FAIL' : 'INVESTIGATE';
    results.push({ query, precision: p, recall: r, status, retrieved: candidates.slice(0, 5), expected, forbiddenHits });
    const icon = status === 'PASS' ? 'PASS' : status === 'FAIL' ? 'FAIL' : 'INV';
    console.log(`[${icon}] P=${p} R=${r} -- ${query.slice(0, 60)}`);
  }

  const passes = results.filter(r => r.status === 'PASS').length;
  const fails = results.filter(r => r.status === 'FAIL').length;

  const dateStr = new Date().toISOString().slice(0, 10);
  const reportDir = join(projectPath, `_outputs/validation/${dateStr}`);
  mkdirSync(reportDir, { recursive: true });

  const reportLines = [
    `# Validation Report -- ${dateStr}\n\n`,
    `**Results: ${passes}/${results.length} pass, ${fails} fail**\n\n`,
    '| Status | P | R | Query |\n',
    '|---|---|---|---|\n',
  ];
  for (const r of results) {
    reportLines.push(`| ${r.status} | ${r.precision} | ${r.recall} | \`${r.query.slice(0, 70)}\` |\n`);
  }
  reportLines.push('\n## Detail\n\n');
  for (const r of results) {
    reportLines.push(`**${r.query}**  \nExpected: ${JSON.stringify(r.expected)}  \nRetrieved (top 5): ${JSON.stringify(r.retrieved)}  \nForbidden hits: ${JSON.stringify(r.forbiddenHits)}\n\n`);
  }

  const reportPath = join(reportDir, 'REPORT.md');
  writeFileSync(reportPath, reportLines.join(''));
  console.log(`\nReport: ${reportPath}`);

  if (fails > 0) {
    console.log(`\n${fails} FAIL(s) -- PAUSE-AND-SURFACE TRIGGER if any precision or recall < 0.5`);
    process.exit(1);
  }
}

// CLI entry guard. Set CORE_DEBUG_CLI_ENTRY=1 to log both strings if invocation
// silently no-ops (path-normalization, symlinks, OneDrive virtualization, etc.).
const _cliEntryCanonical = (p) => { try { return realpathSync(p); } catch { return p; } };
const _cliEntryArgv1 = _cliEntryCanonical(process.argv[1]);
const _cliEntrySelf = _cliEntryCanonical(fileURLToPath(import.meta.url));
if (process.env.CORE_DEBUG_CLI_ENTRY) {
  process.stderr.write(`[cli-entry] argv[1]=${JSON.stringify(_cliEntryArgv1)}\n[cli-entry] self  =${JSON.stringify(_cliEntrySelf)}\n[cli-entry] match=${_cliEntryArgv1 === _cliEntrySelf}\n`);
}
if (_cliEntryArgv1 === _cliEntrySelf) {
  main(process.argv.slice(2));
}
